[WPE] Consider actual available size instead of whole screen size for screen.availHeight and screen.availWidth

lauromoura opened the pull request:
#### eeb1b7547fc469a63421727e9689a6053978545c
<pre>
[WPE] Consider actual available size instead of whole screen size for screen.availHeight and screen.availWidth
<a href="https://bugs.webkit.org/show_bug.cgi?id=278511">https://bugs.webkit.org/show_bug.cgi?id=278511</a>

Reviewed by NOBODY (OOPS!).

Add new wpe_monitor_get_available_width/height API to report the
available screen size. This is used to feed JS&apos;s
screen.availWidth/availHeight[1]. If not possible to detect the available
dimensions, fallback to the previous behavior of using the whole screen
size, as it&apos;s also one of the accepted values in the spec:

&gt; The Web-exposed available screen area is one of the following:
&gt; The available area of the rendering surface of the output device, in CSS pixels.
&gt; The area of the output device, in CSS pixels.

<a href="https://drafts.csswg.org/cssom-view/#dom-screen-availw">https://drafts.csswg.org/cssom-view/#dom-screen-availw</a>

In Wayland, use the `xdg_toplevel::configure_bounds` event to get the
information from the compositor. Given Wayland design decisions, we
can&apos;t get the actual screen position of the client, so we default to
(0,0).

This is not a problem regarding the `availWidth/availHeight`,
which is the information settled currently in the CSS spec. For the
record, this position limitation might become an issue when Window
Management&apos;s `availLeft/availTop`[2] eventually become an accepted
standard. This limitation is described in the Window Management github,
in [3].

[1] <a href="https://developer.mozilla.org/en-US/docs/Web/API/Screen/availHeight">https://developer.mozilla.org/en-US/docs/Web/API/Screen/availHeight</a>
[2] <a href="https://w3c.github.io/window-management/#ref-for-dom-screendetailed-availleft">https://w3c.github.io/window-management/#ref-for-dom-screendetailed-availleft</a>
[3] <a href="https://github.com/w3c/window-management/issues/68">https://github.com/w3c/window-management/issues/68</a>

* Source/WebKit/UIProcess/wpe/ScreenManagerWPE.cpp:
(WebKit::ScreenManager::collectScreenProperties const):
* Source/WebKit/WPEPlatform/wpe/WPEMonitor.cpp:
(wpeMonitorGetProperty):
(wpe_monitor_class_init):
(wpe_monitor_get_available_width):
(wpe_monitor_get_available_height):
(wpe_monitor_set_available_size):
* Source/WebKit/WPEPlatform/wpe/WPEMonitor.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEToplevelWayland.cpp:
(wpeToplevelWaylandConstructed):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eeb1b7547fc469a63421727e9689a6053978545c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63699 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43055 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16296 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67720 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14307 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14587 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51319 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9886 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66768 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40014 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55138 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31951 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36694 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13180 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58664 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12842 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69416 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7646 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12407 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58598 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7679 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55231 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58820 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6356 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38876 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39955 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41067 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39698 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEToplevelWayland.cpp:
```diff
@@ -567,8 +570,6 @@ static void wpeToplevelWaylandConstructed(GObject *object)
         zwp_linux_dmabuf_feedback_v1_add_listener(priv->dmabufFeedback, &linuxDMABufFeedbackListener, object);
     }
 
-    wl_display_roundtrip(wpe_display_wayland_get_wl_display(display));
-
```
Why do you need to move this? I think this was needed here before trying to get the current monitor.

lauromoura replied:
I needed to change because doing the roundtrip was actually triggering getting the display information in `ScreenManagerWPE::collectScreenProperties()`, and changing the monitor data in a second `configure_bounds` was not re-updating the collected data.

Maybe when changing monitor properties `ScreenManager` needs to collect the data again? (Currently it listens only to display's `monitor-added`/`monitor-removed` in `ScreenManager`'s constructor for updating)

carlosgcampos replied:
I'm still worried about this change, I think it can break setting the initial device scale factor.

carlosgcampos replied:
I've checked this is not a problem.

carlosgcampos commented on Source/WebKit/UIProcess/wpe/ScreenManagerWPE.cpp:
```diff
@@ -81,7 +81,7 @@ ScreenProperties ScreenManager::collectScreenProperties() const
         ScreenData data;
         data.screenRect = FloatRect(wpe_monitor_get_x(monitor), wpe_monitor_get_y(monitor), width, height);
-        data.screenAvailableRect = data.screenRect;
+        data.screenAvailableRect = FloatRect(0, 0, wpe_monitor_get_available_width(monitor), wpe_monitor_get_available_height(monitor));
```
Well, this is not exactly the available rect, if you have a top panel, which is common, the available rect doesn't start at 0,0. It's true that the screen rect isn't the available rect either, but it's probably less confusing. We also need to handle the case of available width not being supported, because we are getting -1. I could make wpe monitor automatically fallback to the screen size, or do it here.

lauromoura replied:
Yes, the `configure_bounds` for example, only cares about the size, not the position, so does JS's `screen` object (that originated this issue). There's an experimental property for `screen.availTop/Left`, but only blink-based browsers implement so far: https://developer.mozilla.org/en-US/docs/Web/API/ScreenDetailed/availLeft

And there are open points of how getting the position would even possible under wayland. E.g:

https://github.com/w3c/window-management/issues/68
https://github.com/electron/electron/issues/40886

So I think we could settle with the width/height as they're already part of the CSS spec.

And yes, I can change to fall back to the screen dimensions, if we can't determine the available dimension. It's even the second suggested item in the spec: https://drafts.csswg.org/cssom-view/#web-exposed-available-screen-area

> The available area of the rendering surface of the output device, in [CSS pixels](https://drafts.csswg.org/css-values-4/#px).
>  The area of the output device, in [CSS pixels](https://drafts.csswg.org/css-values-4/#px).
>  The area of the [viewport](https://www.w3.org/TR/CSS21/visuren.html#x1), in [CSS pixels](https://drafts.csswg.org/css-values-4/#px).

carlosgcampos reviewed (commented):
I'm not sure about this, tbh. The WebDriver test passes in gtk because it's run under x11 I guess (xvfb). I think it's ok if that particular test doesn't pass under wayland, because wayland doesn't really support the work area like x11.

lauromoura committed:
[WPE] Consider actual available size instead of whole screen size for screen.availHeight and screen.availWidth
https://bugs.webkit.org/show_bug.cgi?id=278511

Reviewed by NOBODY (OOPS!).

Add new wpe_monitor_get_available_width/height API to report the
available screen size. This is used to feed JS's
screen.availWidth/availHeight[1]. If not possible to detect the available
dimensions, fallback to the previous behavior of using the whole screen
size, as it's also one of the accepted values in the spec:

> The Web-exposed available screen area is one of the following:
> The available area of the rendering surface of the output device, in CSS pixels.
> The area of the output device, in CSS pixels.

https://drafts.csswg.org/cssom-view/#dom-screen-availw

In Wayland, use the `xdg_toplevel::configure_bounds` event to get the
information from the compositor. Given Wayland design decisions, we
can't get the actual screen position of the client, so we default to
(0,0).

This is not a problem regarding the `availWidth/availHeight`,
which is the information settled currently in the CSS spec. For the
record, this position limitation might become an issue when Window
Management's `availLeft/availTop`[2] eventually become an accepted
standard. This limitation is described in the Window Management github,
in [3].

[1] https://developer.mozilla.org/en-US/docs/Web/API/Screen/availHeight
[2] https://w3c.github.io/window-management/#ref-for-dom-screendetailed-availleft
[3] https://github.com/w3c/window-management/issues/68

* Source/WebKit/UIProcess/wpe/ScreenManagerWPE.cpp:
(WebKit::ScreenManager::collectScreenProperties const):
* Source/WebKit/WPEPlatform/wpe/WPEMonitor.cpp:
(wpeMonitorGetProperty):
(wpe_monitor_class_init):
(wpe_monitor_get_available_width):
(wpe_monitor_get_available_height):
(wpe_monitor_set_available_size):
* Source/WebKit/WPEPlatform/wpe/WPEMonitor.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEToplevelWayland.cpp:
(wpeToplevelWaylandConstructed):

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEMonitor.cpp:
```diff
@@ -398,6 +440,64 @@ void wpe_monitor_set_size(WPEMonitor* monitor, int width, int height)
+    g_return_if_fail(WPE_IS_MONITOR(monitor));
+    g_return_if_fail(width == -1 || width >= 0);
+    g_return_if_fail(height == -1 || height >= 0);
+
```
Add g_object_freeze_notify/g_object_thaw_notify to make sure when notify is emitted for eny of the properties, botah have been updated

carlosgcampos commented on Source/WebKit/UIProcess/wpe/ScreenManagerWPE.cpp:
```diff
@@ -81,7 +81,9 @@ ScreenProperties ScreenManager::collectScreenProperties() const
         ScreenData data;
         data.screenRect = FloatRect(wpe_monitor_get_x(monitor), wpe_monitor_get_y(monitor), width, height);
-        data.screenAvailableRect = data.screenRect;
+        // TODO: add some API to get the available position to support screen.availTop/Left if it become a standard in the future
```
```suggestion
        // FIXME: add some API to get the available position to support screen.availTop/Left if it become a standard in the future
```
